api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,629 @@
+			providers = append(providers, newProvider)
+		}
+	}
+
+	// check if we need to scale down
+	unusedProviders := []*models.ResourceProvider{}
+	for _, provider := range providers {
+		if !provider.InUse {
+			unusedProviders = append(unusedProviders, provider)
+		}
+	}
+
+	desiredScale := len(providers) - len(unusedProviders)
+
+	actualScale, err := e.scaleTo(clusterName, desiredScale, unusedProviders)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	info := &models.ScalerRunInfo{
+		EnvironmentID:           clusterName,
+		PendingResources:        resourceConsumerModels(consumers),
+		ResourceProviders:       resourceProviderModels(providers),
+		ScaleBeforeRun:          scaleBeforeRun,
+		DesiredScaleAfterRun:    desiredScale,
+		ActualScaleAfterRun:     actualScale,
+		UnusedResourceProviders: len(unusedProviders),
+	}
+
+	return info, errors.MultiError(errs)
+}
+
+func resourceConsumerModels(consumers []models.ResourceConsumer) []models.ResourceConsumer {
+	consumerModels := make([]models.ResourceConsumer, len(consumers))
+	for i, consumer := range consumers {
+		newConsumer := models.ResourceConsumer{
+			ID:     consumer.ID,
+			Memory: consumer.Memory,
+			Ports:  consumer.Ports,
+		}
+		consumerModels[i] = newConsumer
+	}
+
+	return consumerModels
+}
+
+func resourceProviderModels(providers []*models.ResourceProvider) []models.ResourceProvider {
+	providerModels := make([]models.ResourceProvider, len(providers))
+	for i, provider := range providers {
+		newProvider := models.ResourceProvider{
+			ID:              provider.ID,
+			InUse:           provider.InUse,
+			UsedPorts:       provider.UsedPorts,
+			AvailableMemory: provider.AvailableMemory,
+		}
+		providerModels[i] = newProvider
+	}
+
+	return providerModels
+}
+
+func (e *EnvironmentScaler) getResourceProviders(clusterName string) ([]*models.ResourceProvider, error) {
+	listContainerInstancesInput := &ecs.ListContainerInstancesInput{}
+	listContainerInstancesInput.SetCluster(clusterName)
+
+	containerInstanceARNs := []*string{}
+	listContainerInstancesPagesFN := func(output *ecs.ListContainerInstancesOutput, lastPage bool) bool {
+		containerInstanceARNs = append(containerInstanceARNs, output.ContainerInstanceArns...)
+
+		return !lastPage
+	}
+
+	if err := e.Client.ECS.ListContainerInstancesPages(listContainerInstancesInput, listContainerInstancesPagesFN); err != nil {
+		return nil, err
+	}
+
+	describeContainerInstancesInput := &ecs.DescribeContainerInstancesInput{}
+	describeContainerInstancesInput.SetCluster(clusterName)
+	describeContainerInstancesInput.SetContainerInstances(containerInstanceARNs)
+
+	output, err := e.Client.ECS.DescribeContainerInstances(describeContainerInstancesInput)
+	if err != nil {
+		return nil, err
+	}
+
+	result := []*models.ResourceProvider{}
+	if len(output.ContainerInstances) == 0 {
+		return result, nil
+	}
+
+	for _, instance := range output.ContainerInstances {
+		if !aws.BoolValue(instance.AgentConnected) {
```

Comment:
As discussed in slack - there is an incorrect assumption being made here. Either remove this (as we don't handle this case when container instance count == min cluster count. Or would need to check the connection state for a duration. I.e., if disconnected for more than 3 minutes then consider it permanently disconnected.

In practice - checking for a certain duration will add considerable complexity which I don't think is warranted given:

1. we don't do anything even if the agent is disconnected when current cluster count == min cluster count
2. how often does this situation actually happen?